internal/provider/resource_user.go

```diff
@@ -5,143 +5,293 @@ import (
+// and planned state values should be read from the
+// CreateRequest and new state values set on the CreateResponse.
+func (r *userResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if r.client == nil {
+		resp.Diagnostics.AddError(
+			"Unconfigured HTTP Client",
+			"Expected configured HTTP client. Please report this issue to the provider developers.",
+		)
 
-	var diags diag.Diagnostics
+		return
+	}
 
-	name := d.Get("name").(string)
-	displayName := d.Get("display_name").(string)
-	password := d.Get("password").(string)
+	var data userResourceModel
 
-	_, err := client.Users.CreateUser(juju.CreateUserInput{
-		Name:        name,
-		DisplayName: displayName,
-		Password:    password,
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	_, err := r.client.Users.CreateUser(juju.CreateUserInput{
+		Name:        data.Name.ValueString(),
+		DisplayName: data.DisplayName.ValueString(),
+		Password:    data.Password.ValueString(),
 	})
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create user resource, got error: %s", err))
+		return
 	}
+	tflog.Trace(ctx, fmt.Sprintf("created user resource %q", data.Name))
 
-	d.SetId(fmt.Sprintf("user:%s", name))
-
-	return diags
+	// Save data into Terraform state
+	data.ID = types.StringValue(newIDFromUserName(data.Name.ValueString()))
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
 }
 
-func resourceUserRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+// Read is called when the provider must read resource values in order
+// to update state. Planned state values should be read from the
+// ReadRequest and new state values set on the ReadResponse.
+// Take the juju api input from the ID, it may not exist in the plan.
+// Only set optional values if they exist.
+func (r *userResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Prevent panic if the provider has not been configured.
+	if r.client == nil {
+		resp.Diagnostics.AddError(
+			"Unconfigured HTTP Client",
+			"Expected configured HTTP client. Please report this issue to the provider developers.",
+		)
+		return
+	}
 
-	var diags diag.Diagnostics
+	var data userResourceModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	id := strings.Split(d.Id(), ":")
-	name := id[1]
-	response, err := client.Users.ReadUser(name)
+	// If applicable, this is a great opportunity to initialize any necessary
+	// provider client data and make a call using it.
+	response, err := r.client.Users.ReadUser(userNameFromID(data.ID.ValueString()))
```

Comment:
Adding a user to juju and importing it onto the terraform exposed an assumption here. Here's what I did:

`juju add-user qatest`

added the following into the plan:
```tf
resource "juju_user" "those" {
  name         = "qatest"
}
```

and ran `terraform import juju_user.those qatest` and it panics.

The problem is that the ID leaking back and forth between the provider and juju,  the ID coming from the Juju's side is not in the "user:value" format.

```shell
juju_user.those: Importing from ID "qatest"...
juju_user.those: Import prepared!
  Prepared juju_user for import
juju_user.those: Refreshing state... [id=qatest]
╷
│ Error: Plugin did not respond
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadResource call. The plugin logs
│ may contain more details.
╵


Stack trace from the terraform-provider-juju_v0.4.1 plugin:

panic: runtime error: index out of range [1] with length 1

goroutine 146 [running]:
github.com/juju/terraform-provider-juju/internal/provider.userNameFromID(...)
        /home/caner/go/src/github.com/terraform-provider-juju/internal/provider/resource_user.go:296
github.com/juju/terraform-provider-juju/internal/provider.(*userResource).Read(0xc00048a018, {0x284da20, 0xc000a0a4b0}, {{{{0x2856758, 0xc000a0ad80}, {0x20b32c0, 0xc000a0ad20}}, {0x285b750, 0xc0003b8a00}}, 0xc00048a050, ...}, ...)
        /home/caner/go/src/github.com/terraform-provider-juju/internal/provider/resource_user.go:169 +0x66e
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ReadResource(0xc000577760, {0x284da20, 0xc000a0a4b0}, 0xc000a0a540, 0xc000a8f520)
        /home/caner/go/pkg/mod/github.com/hashicorp/terraform-plugin-framework@v1.3.2/internal/fwserver/server_readresource.go:101 +0x677
github.com/hashicorp/terraform-plugin-framework/internal/proto5server.(*Server).ReadResource(0xc000577760, {0x284da20?, 0xc000a0a360?}, 0xc0006e2040)
        /home/caner/go/pkg/mod/github.com/hashicorp/terraform-plugin-framework@v1.3.2/internal/proto5server/server_readresource.go:56 +0x41a
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.muxServer.ReadResource({0xc0007a1c20, 0xc0003cc2e0, 0xc0007a1c50, 0xc0007a1c80, {0xc00013c4c0, 0x2, 0x2}}, {0x284da20?, 0xc0006d80c0?}, 0xc0006e2040)
        /home/caner/go/pkg/mod/github.com/hashicorp/terraform-plugin-mux@v0.11.1/tf5muxserver/mux_server_ReadResource.go:29 +0x139
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadResource(0xc0003e0140, {0x284da20?, 0xc000630180?}, 0xc00063c060)
        /home/caner/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.18.0/tfprotov5/tf5server/server.go:751 +0x4b1
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0x23212e0?, 0xc0003e0140}, {0x284da20, 0xc000630180}, 0xc0002e0000, 0x0)
        /home/caner/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.18.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:386 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001de5a0, {0x2857460, 0xc000602b60}, 0xc00063e000, 0xc0007bbe00, 0x38c0d10, 0x0)
        /home/caner/go/pkg/mod/google.golang.org/grpc@v1.56.1/server.go:1337 +0xdf3
google.golang.org/grpc.(*Server).handleStream(0xc0001de5a0, {0x2857460, 0xc000602b60}, 0xc00063e000, 0x0)
        /home/caner/go/pkg/mod/google.golang.org/grpc@v1.56.1/server.go:1714 +0xa36
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        /home/caner/go/pkg/mod/google.golang.org/grpc@v1.56.1/server.go:959 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/caner/go/pkg/mod/google.golang.org/grpc@v1.56.1/server.go:957 +0x18c

Error: The terraform-provider-juju_v0.4.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

Reply:
`terraform import juju_user.those qatest` is an invalid import line. However this should have a nice failure message not a panic.

Reply:
@hmlanigan why is that an invalid line? I thought the ID field is whatever we see in the `juju users`, no? Is the correct id `user:qatest` ?

Reply:
@cderici you found a panic which needed to be fixed by using incorrect input.  A good test.